### PR TITLE
fix: pq index does not handle dot product metric correctly during search

### PR DESCRIPTION
**Pull request by rok**
See #1508

### CODE REVIEW

File: rust/lance/src/index/vector/pq.rs [hunk 1]
```diff
@@ -131,6 +131,45 @@ impl PQIndex {
...
+        // Distance table: `[f32: num_sub_vectors(row) * num_centroids(column)]`.
+        let mut distance_table: Vec<f32> = vec![];
+
+        let sub_vector_length = self.dimension / self.num_sub_vectors;
+        for i in 0..self.num_sub_vectors {
+            // The sub-vector section of the query vector.
+            let from = key.slice(i * sub_vector_length, sub_vector_length);
+            let subvec_centroids = self.pq.centroids(i);
+            let distances = dot_distance_batch(
+                as_primitive_array::<Float32Type>(&from).values(),
```
Comment (eddyxu) [inline rust/lance/src/index/vector/pq.rs:149]: Lets use the new api `from.as_primitive().values()` here?

File: rust/lance/src/index/vector/pq.rs [hunk 2]
```diff
@@ -267,8 +306,12 @@ impl VectorIndex for PQIndex {
             // Pre-compute distance table for each sub-vector.
             let distances = if this.metric_type == MetricType::L2 {
                 this.fast_l2_distances(&query.key, code.as_ref())?
-            } else {
+            } else if this.metric_type == MetricType::Dot {
```
Comment (eddyxu) [inline rust/lance/src/index/vector/pq.rs:309]: We can just use a `match metric_type` ?

Comment (rok) [reply rust/lance/src/index/vector/pq.rs]: Done.

File: rust/lance/src/index/vector/pq.rs (hunk 1 above)
Comment (rok) [reply rust/lance/src/index/vector/pq.rs]: Turns out `from` is already primitive so changed to `from.values()`.

File: rust/lance/src/index/vector/ivf.rs [hunk 3]
```diff
@@ -1511,4 +1511,50 @@ mod tests {
...
+        assert_eq!(1, results.len());
+        assert_eq!(5, results[0].num_rows());
+
+        for batch in results.iter() {
+            let dist = &batch["_distance"];
+            assert!(dist
+                .as_primitive::<Float32Type>()
+                .values()
+                .iter()
+                .all(|v| (-2.0 * DIM as f32..0.0).contains(v)));
```
Comment (rok) [inline rust/lance/src/index/vector/ivf.rs:1557]: Our dot product always returns negative value and with input value of `[0, 1)` we get a range of `(-1, 0]`. I suppose that is expected?

**Review by eddyxu — APPROVED** [review]

**Pull request merged**